test/coinselector-test.js

```diff
@@ -0,0 +1,335 @@
+          selectedValues += pointers[i].effectiveValue;
+        }
+
+        assert(selectedValues >= target);
+      }
+    });
+  });
+});
+
+const workers = new WorkerPool({
+  enabled: true,
+  size: 2
+});
+
+describe('Integration', function () {
+  this.timeout(1000000);
+
+  const wdb = new WalletDB({workers});
+
+  function fromU32(num) {
+    const data = Buffer.allocUnsafe(4);
+    data.writeUInt32LE(num, 0, true);
+    return data;
+  }
+
+  function nextBlock(wdb) {
+    return fakeBlock(wdb.state.height + 1);
+  }
+
+  function fakeBlock(height) {
+    const prev = hash256.digest(fromU32((height - 1) >>> 0));
+    const hash = hash256.digest(fromU32(height >>> 0));
+    const root = hash256.digest(fromU32((height | 0x80000000) >>> 0));
+
+    return {
+      hash: hash,
+      prevBlock: prev,
+      merkleRoot: root,
+      time: 500000000 + (height * (10 * 60)),
+      bits: 0,
+      nonce: 0,
+      height: height
+    };
+  }
+
+  function dummyInput() {
+    const hash = random.randomBytes(32);
+    return Input.fromOutpoint(new Outpoint(hash, 0));
+  }
+
+  before(async () => {
+    await wdb.open();
+    await workers.open();
+  });
+
+  let oldBalance, newBalance, oldCoins, newCoins;
+
+  after(async () => {
+    console.log('Amount saved in fees :', newBalance - oldBalance);
+    console.log('Coins in UTXO pool using old selection :', oldCoins);
+    console.log('Coins in UTXO pool using new selection :', newCoins);
+    await wdb.close();
+    await workers.close();
+  });
+
+  for (const useSelectEstimate of [true, false]) {
+    it('should send transactions using old selection', async () => {
+      const alice = await wdb.create();
+      const bob = await wdb.create();
+
+      for (let i = 0; i < 6000; i++) {
+        const payment = data[i];
+        let value = Amount.value(payment.value);
+        const rate = Amount.value(payment.rate);
+
+        // remove dust outputs
+        if (Math.abs(value) < 500)
+          continue;
+
+        let tx;
+
+        if (value > 0) {
+          // send to Alice's wallet
+          const t1 = new MTX();
+          t1.addInput(dummyInput());
+          t1.addOutput(await alice.receiveAddress(), value);
+          tx = t1.toTX();
+        } else {
+          // send from Alice's wallet
+          const address = await bob.receiveAddress();
+          value = -value;
+          tx = await alice.send({
+            outputs: [{value, address}],
+            rate,
+            useSelectEstimate
+          });
+        }
+
+        // confirm tx
+        await wdb.addBlock(nextBlock(wdb), [tx]);
+      }
+      // check balance after simulation
+      const balance = await alice.getBalance();
+      if (useSelectEstimate) {
+        oldBalance = balance.unconfirmed;
+        oldCoins = balance.coin;
+      } else {
+        newBalance = balance.unconfirmed;
+        newCoins = balance.coin;
+      }
+    });
+  }
+
+  it('should select mixed input coins', async () => {
```

Comment:
Well, that's a good idea but I don't want to do that because the last test proves that the new selection algorithm is an improvement over the old one. Don't want to do multiple things inside a single test.
This test was written to see if the new selection can select mixed inputs. I still don't get the idea behind adding this test. Coin Selector doesn't ever take into account the input type. It doesn't care if it's P2PKH to P2WPKH. The only thing that is considered when selecting coins is the `effectiveValue`.

Reply:
Mixed input TXs can cause unexpected results, especially if some data is cached between inputs. And there could be a regression if CoinSelector doesn't get effective value correctly (maybe it passes in the wrong multisig account type or something). So there is good reason to test it. Not to mention plenty of bcoin wallets in production have mixed nested / native segwit. I don't want to find out in a month that some service is inadvertently paying huge miner fees.

Reply:
Okay, I got your point.